analyses/cms-open-data-ttbar/ttbar_analysis_pipeline.py

```diff
@@ -182,112 +174,99 @@ def process(self, events):
+                region_selection = selections.all(region)
+                region_jets = jets[region_selection]
+                region_weights = np.ones(len(region_jets)) * xsec_weight
                 if region == "4j1b":
-                    region_filter = ak.sum(selected_jets.btagCSVV2 >= B_TAG_THRESHOLD, axis=1) == 1
-                    selected_jets_region = selected_jets[region_filter]
-                    # use HT (scalar sum of jet pT) as observable
-                    pt_var_modifier = (
-                        events[event_filters][region_filter][pt_var]
-                        if "res" not in pt_var
-                        else events[pt_var][jet_filter][event_filters][region_filter]
-                    )
-                    observable = ak.sum(selected_jets_region.pt * pt_var_modifier, axis=-1)
-
+                    observable = ak.sum(region_jets.pt,axis=-1)
                 elif region == "4j2b":
-                    region_filter = ak.sum(selected_jets.btagCSVV2 > B_TAG_THRESHOLD, axis=1) >= 2
-                    selected_jets_region = selected_jets[region_filter]
-
                     # reconstruct hadronic top as bjj system with largest pT
-                    # the jet energy scale / resolution effect is not propagated to this observable at the moment
-                    trijet = ak.combinations(selected_jets_region, 3, fields=["j1", "j2", "j3"])  # trijet candidates
+                    trijet = ak.combinations(region_jets, 3, fields=["j1", "j2", "j3"])  # trijet candidates
                     trijet["p4"] = trijet.j1 + trijet.j2 + trijet.j3  # calculate four-momentum of tri-jet system
                     trijet["max_btag"] = np.maximum(trijet.j1.btagCSVV2, np.maximum(trijet.j2.btagCSVV2, trijet.j3.btagCSVV2))
                     trijet = trijet[trijet.max_btag > B_TAG_THRESHOLD]  # at least one-btag in trijet candidates
                     # pick trijet candidate with largest pT and calculate mass of system
                     trijet_mass = trijet["p4"][ak.argmax(trijet.p4.pt, axis=1, keepdims=True)].mass
                     observable = ak.flatten(trijet_mass)
-
-                ### histogram filling
-                if pt_var == "pt_nominal":
-                    # nominal pT, but including 2-point systematics
-                    histogram.fill(
+                syst_var_name = f"{syst_var}"
+                # Break up the filling into event weight systematics and object variation systematics
+                if syst_var in event_systs:
+                    for i_dir,direction in enumerate(["up","down"]):
+                        # Should be an event weight systematic with an up/down variation
+                        if syst_var.startswith("btag_var"):
+                            i_jet = int(syst_var.rsplit("_",1)[-1])   # Kind of fragile
+                            wgt_variation = self.cset["event_systematics"].evaluate("btag_var",direction,region_jets.pt[:,i_jet])
```

Comment:
One thing I'm wondering about here (not for this PR but maybe the future): by matching the name `btag_var` to the correction that is defined, the main thing here that differs per systematic variation (and means we need some dedicated per-systematic handling of the `.evaluate` call) is the information about which columns to feed in.

If we had an extra intermediate layer that knows which column the correction needs, and can automatically extract the right information, then the section here would perhaps not even need a dedicated treatment per systematic. In that case the `.evaluate` call equivalent would essentially get all the event information and then figure out internally what to pass on to `correctionlib`. I am not sure if this is all that feasible, or even a good idea. Maybe it only makes the code here look more streamlined by hiding the complication away?

cc @nsmith-

Reply:
It certainly could be done automatically, to the extent that the input names available from the correction object can be matched to columns in the file. That could be non-trivial, as it is only by convention.